Uniform structure split

PatrickMassot opened the pull request:
This splits `uniform_space/basic` and disentangle the mess of `topological_structure` and `completion`. In particular, files wanting continuity of multiplication on a topological group won't need the whole theory of uniform spaces...

PatrickMassot committed:
chore(topology): try splitting uniform_space and topological_structure

PatrickMassot committed:
fix(topology/algebra/ring): remove duplicate instance

PatrickMassot committed:
fix(topology/*): fix build?

PatrickMassot committed:
fix(topology): Fix headers

sgouezel commented on src/topology/bounded_continuous_function.lean:
```diff
@@ -215,6 +216,8 @@ a common modulus of continuity and taking values in a compact set forms a compac
 and several useful variations around it. -/
 
+variables (A : set (α →ᵇ β))
+#check (coe_fn_trans : has_coe_to_fun A)
```
Remove

PatrickMassot replied:
thanks.

sgouezel commented on src/topology/bounded_continuous_function.lean:
```diff
@@ -257,6 +260,7 @@ begin
 
   /- Associate to every function a discrete approximation, mapping each point in `tα`
   to a point in `tβ` close to its true image by the function. -/
+  letI := @coe_fn_trans ↥A _ _  (@bounded_continuous_function.has_coe_to_fun α β _ _),
```
This is strange that you need this.

PatrickMassot replied:
This is discussed in https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/uniform.20split Indeed the type class mechanism is behaving strangely. There is also a fix in `exponential.lean` that shouldn't be needed.

sgouezel commented on src/topology/algebra/ring.lean:
```diff
@@ -0,0 +1,111 @@
+instance topological_ring_quotient_topology : topological_space N.quotient :=
+by dunfold ideal.quotient submodule.quotient; apply_instance
+
+-- this should be quotient_add_saturate ...
```
Isn't it a good occasion to rename the lemma?

PatrickMassot replied:
I removed the comment, the current name is clearly better than the suggestion.

PatrickMassot committed:
Minor fixes pointed out by reviewers

PatrickMassot committed:
Fix the instance issue by further split

This improves isolation of completions, which was our of the main goals
of this PR. Not doing that yesterday was an oversight from me.

jcommelin commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -116,25 +116,27 @@ classical.by_cases
+          ... = ∥∥x∥⁻¹∥ * ∥A x∥ * ∥x∥  : by rw[‹∥∥x∥⁻¹∥ = ∥x∥⁻¹›]; ring
           ... = ∥∥x∥⁻¹• A x ∥ * ∥x∥    : by rw[←normed_space.norm_smul ∥x∥⁻¹ (A x)]
-          ... = ∥A (∥x∥⁻¹• x)∥ * ∥x∥   : begin rw[show ⇑A = A.val, from rfl, A.property.smul] end
+          ... = ∥A (∥x∥⁻¹• x)∥ * ∥x∥   : by {
```
Would it make sense to write this as something like
```lean
show  ∥∥x∥⁻¹ • A.val x∥ * ∥x∥ = ∥A.val (∥x∥⁻¹ • x)∥ * ∥x∥, by rw A.property.smul
```
Anyway, I don't really care about golfing too much.

PatrickMassot replied:
Actually I don't understand why this modification was needed. I can only say it has nothing to do with completions, since completions are not imported by this file, even indirectly.

jcommelin commented on src/topology/algebra/group.lean:
```diff
@@ -1,105 +1,241 @@
+{ continuous_to_fun  := continuous_inv',
+  continuous_inv_fun := continuous_inv',
+  .. equiv.inv α }
+attribute [to_additive homeomorph.inv._proof_1] homeomorph.inv._proof_1
```
Shouldn't this be `to_additive homeomorph.neg._proof_1`? Idem on the next two lines.

PatrickMassot replied:
I think so (although this has nothing to do with this PR)

jcommelin commented on src/topology/algebra/group_completion.lean:
```diff
@@ -0,0 +1,111 @@
+is_add_group_hom_extension (hf.comp (continuous_coe _))
+
+section instance_max_depth
+-- TODO: continuous_add requires some long proofs through
```
This TODO is somewhat cryptic. I guess it should either be expanded or removed.

PatrickMassot replied:
This was written by @johoelzl so I don't dare modifying it

jcommelin commented on src/topology/algebra/group_completion.lean:
```diff
@@ -0,0 +1,111 @@
+-- TODO: continuous_add requires some long proofs through
+-- uniform_add_group / topological_add_group w.r.t prod / completion etc
+set_option class.instance_max_depth 52
+-- set_option trace.class_instances true
```
This line can be removed.

jcommelin commented on src/topology/algebra/ring.lean:
```diff
@@ -0,0 +1,101 @@
+
+end topological_comm_ring
+
+
```
```suggestion
```

jcommelin commented on src/topology/algebra/uniform_group.lean:
```diff
@@ -0,0 +1,503 @@
+
+end uniform_add_group
+
+
```
Remove some empty lines

jcommelin commented on src/topology/algebra/uniform_group.lean:
```diff
@@ -0,0 +1,503 @@
+
+
+
+/- open filter
```
Should this commented code go in mathlib? I guess it can be removed.

jcommelin commented on src/topology/algebra/uniform_group.lean:
```diff
@@ -0,0 +1,503 @@
+variables {α : Type*} {β : Type*} {γ : Type*} {δ : Type*}
+variables [add_comm_group α] [add_comm_group β] [add_comm_group γ]
+
+/- TODO: when modules are changed to have more explicit base ring, then change replace `is_Z_bilin`
```
Modules now have a more explicit base ring. I don't care if we fix this TODO during this refactor. We can also make a mental not to fix it after this refactor is merged.

PatrickMassot replied:
Yes, I'd like to keep this a pure reorganization PR.

jcommelin reviewed (commented):
Some minor comments

PatrickMassot committed:
fix(topology/algebra): stylistic fixes following Johan
